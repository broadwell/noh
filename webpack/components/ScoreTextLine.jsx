/* eslint no-param-reassign: ["error", { "props": false }] */

import React from "react";
import PropTypes from "prop-types";
import CellText from "./CellText";
import { fillGrid } from "../utils";

const ScoreTextLine = props => {
  if (props.textGrid.length === 0) {
    return <CellText text="" length={props.length} />;
  }
  for (let i = 0; i < props.textGrid.length; i += 1) {
    if (props.textGrid[i].voices && props.textGrid[i].voices[0]) {
      if (i === 0) {
        props.textGrid[i].beginning = true;
      }
      if (
        typeof props.textGrid[i - 1] !== "undefined" &&
        props.textGrid[i - 1].voices &&
        (props.textGrid[i - 1].voices.length === 0 ||
          props.textGrid[i].voices[0] !== props.textGrid[i - 1].voices[0])
      ) {
        props.textGrid[i].beginning = true;
      }
      if (
        typeof props.textGrid[i + 1] !== "undefined" &&
        (props.textGrid[i + 1].voices === undefined ||
          props.textGrid[i + 1].voices.length === 0)
      ) {
        props.textGrid[i].end = true;
      }
      if (
        typeof props.textGrid[i + 1] !== "undefined" &&
        props.textGrid[i + 1].voices &&
        props.textGrid[i + 1].voices[0] &&
        props.textGrid[i].voices[0] !== props.textGrid[i + 1].voices[0]
      ) {
        props.textGrid[i].end = true;
      }
      if (typeof props.textGrid[i + 1] === "undefined") {
        props.textGrid[i].end = true;
      }
    }
  }
  for (let i = 0; i < props.textGrid.length; i += 1) {
    for (let j = 0; j < props.rangeGrid.length; j += 1) {
      if (
        props.textGrid[i].start >= props.rangeGrid[j].start &&
        props.textGrid[i].start <
          props.rangeGrid[j].start + props.rangeGrid[j].length
      ) {
        props.textGrid[i].vocalRange = props.rangeGrid[j].text;
      }
    }
  }
  const fullData = fillGrid(props.textGrid, props.length);
  const textCells = fullData.map((cell, idx) => (
    <CellText
      text={cell.text}
      length={cell.length}
      key={`textCell${idx}`} // eslint-disable-line react/no-array-index-key
      vocalRange={cell.vocalRange}
      voiceType={cell.voices && cell.voices[0] ? cell.voices[0] : ""}
      beginning={cell.beginning || false}
      end={cell.end || false}
    />
  ));
  return textCells;
};

ScoreTextLine.propTypes = {
  textGrid: PropTypes.arrayOf(
    PropTypes.shape({
      voices: PropTypes.arrayOf(PropTypes.string),
      beginning: PropTypes.bool,
      end: PropTypes.bool,
      start: PropTypes.number,
      vocalRange: PropTypes.string
    })
  ).isRequired,
  length: PropTypes.number.isRequired,
  rangeGrid: PropTypes.arrayOf(
    PropTypes.shape({
      start: PropTypes.number,
      text: PropTypes.string,
      length: PropTypes.number
    })
  ).isRequired
};

export default ScoreTextLine;
