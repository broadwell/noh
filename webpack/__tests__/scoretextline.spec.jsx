import React from "react";
import { shallow } from "enzyme";
import ScoreTextLine from "../components/ScoreTextLine";
import phrases from "./__fixtures__/phrases.json";

describe("<ScoreTextLine>", () => {
  it("renders as expected by default", () => {
    const component = shallow(
      <ScoreTextLine
        textGrid={phrases.phrases[0].syllableText.grid}
        numberGrid={phrases.phrases[0].syllableNumber.grid}
        length={13}
        rangeGrid={phrases.phrases[0].vocalRange.grid}
      />
    );
    expect(component).toMatchSnapshot();
  });
  it("renders as expected with empty text and number grid", () => {
    const component = shallow(
      <ScoreTextLine
        textGrid={[]}
        numberGrid={[]}
        length={13}
        rangeGrid={phrases.phrases[0].vocalRange.grid}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
