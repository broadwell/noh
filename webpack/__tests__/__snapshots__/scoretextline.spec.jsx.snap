// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScoreTextLine> renders as expected by default 1`] = `
Array [
  <CellText
    key="textCell0"
    length={1}
    text="ma"
    textIsCongruent={true}
    vocalRange="l"
    voiceType="jiutai"
  />,
  <CellText
    key="textCell1"
    length={1}
    text="ta"
    textIsCongruent={true}
    vocalRange="l"
    voiceType=""
  />,
  <CellText
    key="textCell2"
    length={1}
    text="wa"
    textIsCongruent={true}
    vocalRange="l"
    voiceType="jiutai"
  />,
  <CellText
    key="textCell3"
    length={1}
    text="ga"
    textIsCongruent={true}
    vocalRange="l"
    voiceType="shite"
  />,
  <CellText
    key="textCell4"
    length={1}
    text="cho"
    textIsCongruent={true}
    vocalRange="l"
    voiceType=""
  />,
  <CellText
    key="textCell5"
    length={1}
    text="o"
    textIsCongruent={true}
    vocalRange="l"
    voiceType=""
  />,
  <CellText
    key="textCell6"
    length={1}
    text="no"
    textIsCongruent={true}
    vocalRange="l"
    voiceType="jiutai"
  />,
  <CellText
    key="textCell7"
    length={1}
    text="so"
    textIsCongruent={true}
    vocalRange="l"
    voiceType=""
  />,
  <CellText
    key="textCell8"
    length={1}
    text="no"
    textIsCongruent={true}
    vocalRange="l"
    voiceType=""
  />,
  <CellText
    key="textCell9"
    length={1}
    text="ha"
    textIsCongruent={true}
    vocalRange="l"
    voiceType=""
  />,
  <CellText
    key="textCell10"
    length={1}
    text="ji"
    textIsCongruent={true}
    vocalRange="l"
    voiceType=""
  />,
  <CellText
    key="textCell11"
    length={1}
    text="me"
    textIsCongruent={true}
    vocalRange="l"
    voiceType=""
  />,
  <CellText
    key="textCell12"
    length={1}
    text="(e)"
    textIsCongruent={true}
    vocalRange="l"
    voiceType="shite"
  />,
]
`;

exports[`<ScoreTextLine> renders as expected by default 2`] = `
<CellText
  length={13}
  text=""
  textIsCongruent={false}
  vocalRange=""
  voiceType=""
/>
`;
